Mathlib/LinearAlgebra/QuadraticForm/Basis.lean

```diff
@@ -18,7 +20,87 @@ open LinearMap (BilinMap)
 
 namespace QuadraticMap
 
-variable {ι R M N} [LinearOrder ι]
+section
+
+variable {ι R M N}
+
+variable [CommRing R] [AddCommGroup M] [AddCommGroup N] [Module R M] [Module R N]
+
+open Finsupp in
+theorem map_finsuppSum' (Q : QuadraticMap R M N) (f : ι →₀ R) (g : ι → R → M) :
+    Q (f.sum g) =
+      ∑ p ∈ f.support.sym2, polarSym2 Q (p.map fun i => g i (f i))
+        - ∑ i ∈ f.support, Q (g i (f i)) := by
+  exact Q.map_sum' _ (fun i => g i (f i))
+
+lemma polarSym2_map_smul_apply (Q : QuadraticMap R M N) (g : ι → M) (l : ι → R) (p : Sym2 ι) :
+    polarSym2 Q (p.map (l • g)) = (p.map l).mul • polarSym2 Q (p.map g) := by
+  obtain ⟨_,_⟩ := p
+  simp only [Pi.smul_apply', Sym2.map_pair_eq, polarSym2_sym2Mk, polar_smul_right, polar_smul_left,
+    Function.comp_apply, Sym2.mul_mk, ← smul_assoc, smul_eq_mul, mul_comm]
+
+open Finsupp in
+theorem apply_linearCombination' (Q : QuadraticMap R M N) {g : ι → M} (l : ι →₀ R) :
+    Q (linearCombination R g l) =
+      linearCombination R (polarSym2 Q ∘ Sym2.map g) (Sym2.mulFinsupp l) -
+      linearCombination R (Q ∘ g) (l * l)  := by
+  simp_rw [linearCombination_apply, map_finsuppSum', map_smul, mul_smul]
+  rw [Finsupp.sum_of_support_subset (l * l) (support_mul_subset_left (g₁ := l))
+    (fun i a => a • (⇑Q ∘ g) i) (fun _ _=> by simp only [Function.comp_apply, zero_smul])]
+  simp only [mul_apply, Function.comp_apply]
+  simp only [←smul_eq_mul, smul_assoc, sub_left_inj]
+  rw [Finsupp.sum_of_support_subset (Sym2.mulFinsupp l) (Sym2.support_mulFinsupp_subset l) _
+    (fun p hp => zero_smul R ((polarSym2 Q) (Sym2.map g p)))]
+  apply Finset.sum_congr rfl
+  intro _ _
+  rw [Sym2.mulFinsupp_eq_mul_comp_map, Function.comp_apply, ← polarSym2_map_smul_apply]
+  rfl
+
+open Finsupp in
+theorem sum_polar_sub_repr_sq (Q : QuadraticMap R M N) (bm : Basis ι R M) (x : M) :
+    linearCombination R (polarSym2 Q ∘ Sym2.map bm) (Sym2.mulFinsupp (bm.repr x)) -
+      linearCombination R (Q ∘ bm) ((bm.repr x) * (bm.repr x)) = Q x := by
+  rw [← apply_linearCombination', Basis.linearCombination_repr]
+
+variable [DecidableEq ι]
+
+-- c.f. `_root_.map_finsupp_sum`
+open Finsupp in
+theorem map_finsuppSum (Q : QuadraticMap R M N) (f : ι →₀ R) (g : ι → R → M) :
+    Q (f.sum g) =
+      f.sum (fun i r => Q (g i r)) +
+      ∑ p ∈ f.support.sym2 with ¬ p.IsDiag,
+        polarSym2 Q (p.map fun i => g i (f i))  := by
```

Comment:
and if you also define `Finsupp.sym2OffDiag`, you can restate this as
```suggestion
      f.sym2.sum fun p a =>
        polarSym2 Q (p.map fun i => g i (f i))  := by
```
or something like this?

Reply:
So start with something like:
```
/--
Off-diagonal multiplication
-/
def mulOffDiag : R → R → R := fun a b => if a = b then (0 : R) else a * b

/--
Multiplication as a function from `Sym2`.
-/
def Sym2.mulOffDiag : Sym2 R → R := lift ⟨_root_.mulOffDiag, fun a b => by
  rw [_root_.mulOffDiag, _root_.mulOffDiag, mul_comm]
  simp_rw [eq_comm]⟩
```
?

Reply:
I think this will add `DecidableEq R` as a hypothesis to `QuadraticMap.map_finsuppSum` though, so slightly less general than what we had before?

Reply:
Then I think you'd need to set this up more generally for `F : M × M -> M` such that `F a b = F b a`, `F 0 a = 0` and `F b 0 = 0` in order to do something similar for `polarSym2`  in `QuadraticMap.map_sum`?

Reply:
Perhaps something for a follow up PR?

Reply:
@YaelDillies I think I went off down the wrong path multiplying things together.

My thinking got to #22313, but I can't see how to use that in the above statement without either additional hypothesis on `g`.

Reply:
Yeah sorry actually I think I completely hallucinated the solution here. Feel free to revert to the last working state

Reply:
Done - not sure why it's failing now:
```
Run scripts/lean-pr-testing-comments.sh lean
##[debug]/usr/bin/bash -e /home/lean/actions-runner/_work/_temp/400c7514-b7de-4731-bb00-f21554e22f6b.sh
/home/lean/actions-runner/_work/_temp/400c7514-b7de-4731-bb00-f21554e22f6b.sh: line 1: scripts/lean-pr-testing-comments.sh: No such file or directory
Error: Process completed with exit code 127.
```

Reply:
Not sure, no. Can you complain on Zulip?

Reply:
Seems to be fixed now.